lib/rrd.ml

```diff
@@ -692,55 +692,72 @@ let xml_to_output rrd output =
-        ",max:"^(Utils.f_to_s ds.ds_max)^",last_ds:0.0,value:0.0,unknown_sec:0}") ds_list))^"]"
-  in
-
-  let do_rra_cdps cdp_list =
-    "ds:["^(String.concat "," (List.map (fun cdp ->
-        "{primary_value:0.0,secondary_value:0.0,value:"^(Utils.f_to_s cdp.cdp_value)^
-        ",unknown_datapoints:"^(Printf.sprintf "%d" cdp.cdp_unknown_pdps)^"}") cdp_list))^"]"
-  in
-
-  let do_database rings =
-    if Array.length rings = 0 then "[]" else
+module Json = struct
+
+  let string fmt = Printf.ksprintf (fun msg -> `String msg) fmt
+  let float x    = string "%.2f" x
+  let record xs  = `O xs
+  let array  xs  = `A xs
+
+  let ty = function
+    | Gauge     -> string "GAUGE"
+    | Absolute  -> string "ABSOLUTE"
+    | Derive    -> string "DERIVE"
+
+  let datasource ds =
+    record
+      [ "name"               , string "%s" ds.ds_name
+      ; "type"               , ty ds.ds_ty
+      ; "minimal_hearbeat"   , string "%s" (Utils.f_to_s ds.ds_mrhb)
+      ; "min"                , string "%s" (Utils.f_to_s ds.ds_min)
+      ; "max"                , string "%s" (Utils.f_to_s ds.ds_max)
+      ; "last_ds"            , float 0.0
+      ; "value"              , float 0.0
+      ; "unknown_sec"        , float 0.0
+      ]
+
+  let cdp x =
+    record
+      [ "primary_value"       , float 0.0
+      ; "secondary_value"     , float 0.0
+      ; "value"               , string "%s" (Utils.f_to_s x.cdp_value)
+      ; "unknown_datapoints"  , float @@ float_of_int @@ x.cdp_unknown_pdps
+      ]
+
+  let get rings rows row col =
+    Fring.peek rings.(col) (rows-row-1) |> Utils.f_to_s |> string "%s"
+
+  let database = function
+    | [||]  -> array []
+    | rings ->
       let rows = Fring.length rings.(0) in
       let cols = Array.length rings in
-      "["^(String.concat ","
-             (Array.to_list
-                (Array.init rows (fun row ->
-                     "["^(String.concat ","
-                            (Array.to_list
-                               (Array.init cols
-                                  (fun col ->
-                                     Utils.f_to_s
-                                       (Fring.peek rings.(col)
-                                          (rows-row-1))))))^"]"))))^"]"
-  in
-
-  let do_rras rra_list =
-    "rra:[{" ^ (String.concat "},{" (List.map (fun rra ->
-        "cf:\""^(cf_type_to_string rra.rra_cf)^
-        "\",pdp_per_row:"^(string_of_int rra.rra_pdp_cnt)^",params:{xff:"^(Utils.f_to_s rra.rra_xff)^"},cdp_prep:{"^(do_rra_cdps (Array.to_list rra.rra_cdps))^
-        "},database:"^(do_database rra.rra_data)) rra_list))^"}]"
-  in
+      array @@ Array.to_list @@ Array.init rows (fun row ->
+          array @@ Array.to_list @@ Array.init cols (fun col -> get rings rows row col))
+
+  let rra x =
+    record
+      [ "cf"           , string "%s" (cf_type_to_string x.rra_cf)
+      ; "pdp_per_row"  , string "%d" x.rra_pdp_cnt
+      ; "params"       , string "%s" (Utils.f_to_s x.rra_xff)
```

Comment:
Previously this was more deeply nested as `params: {xff: rra_xff_value}`, was this changed on purpose?

Reply:
This wasn't on purpose. Will take look.